Stage react auth

flysputnik17 opened the pull request.

flysputnik17 committed:
auth and newsCards API

flysputnik17 committed:
news add refactoring

flysputnik17 committed:
add and getNews working problem with delete news

flysputnik17 committed:
savedNews keywords and delete news fix

flysputnik17 committed:
minor var fix and refactor

flysputnik17 committed:
user login and news render fix

flysputnik17 committed:
keyword and saved Articles per user fix

flysputnik17 committed:
login and register errors messages

flysputnik17 committed:
Email used error fix

flysputnik17 committed:
remote server deploy

flysputnik17 committed:
proxy server

flysputnik17 committed:
README UPDATE

flysputnik17 committed:
username -> name field change

flysputnik17 committed:
auth name field update

flysputnik17 committed:
getNews api update

flysputnik17 committed:
Protected routes and minor fix

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Pavel, 

unfortunately I have to reject the project without reviewing:


* I can't search anything because the page reloads on submit (but it works when I click on the button)
* I can't save articles. See the error https://snipboard.io/ft9YjV.jpg 
*  When an unauthorized user clicks on the icon to save an article, the registration popup window opens.




Please, check yourself according to the checklist (maybe you haven't implemented something)

  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.

flysputnik17 committed:
news search on enter & register popup

flysputnik17 committed:
minor changes

flysputnik17 committed:
add news after login error 400 fix

gennady-bars commented on src/utils/MainApi.js:
```diff
@@ -14,11 +14,11 @@ export default class Auth {
       headers: this.headers,
-      body: JSON.stringify({ email, password, username }),
+      body: JSON.stringify({ email, password, name }),
     }).then(this.checkResponse);
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/components/PopupWithForm/PopupWithForm.css:
```diff
@@ -85,6 +85,15 @@
   margin-top: 9px;
 }
 
```
**Could be improved**

If it’s interesting for you here is a way to make a special wrapper for any `popup` with cross button, `Escape` and `overlay` listeners inside. You should create component `Popup`, store it in `Popup.js`. 

```jsx
import { useEffect } from "react";

const Popup = ({ isOpen, name, onClose, children }) => {
// here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // with this we prevent adding the listener if the popup is not opened
    if (!isOpen) return;
// we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const closeByEscape = (e) => {
      if (e.key === 'Escape') {
        onClose();
      }
    }

    document.addEventListener('keydown', closeByEscape)
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener('keydown', closeByEscape)
// here we watch `isOpen` to add the listener only when it’s opened
}, [isOpen, onClose])

// here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
        onClose();
    }
  }

// then we add the main wrapper with class `popup` and `popup_opened`
  return (
    <div
      className={`popup ${isOpen ? "popup_opened" : ""} popup_type_${name}`}
      onClick={handleOverlay}
    >
    {/* the container for the contents */}
      <div className='popup__container'>
        {/* here will be anything you add as `children`
*/}
        {children}
        {/* add the close button */}
        <button
          className='popup__close'
          type='button'
          onClick={onClose}
        />
      </div>
    </div>
  );
};

export default Popup;


```

And now you can use `Popup` for any popup in the project:  ` ImagePopup`, ` InfoToolTip` and `PopupWithForm`.

```jsx


function PopupWithForm({isOpen, name, onClose, ...props}) {
  return (
    <Popup isOpen={isOpen} name={name} onClose={onClose}>
        <h2 className='popup__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.

gennady-bars commented on src/components/SigninPopup/SigninPopup.jsx:
```diff
@@ -1,35 +1,71 @@
+    setEmail(value);
+
+    // Email validation logic
+    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+    if (!emailRegex.test(value)) {
+      setEmailError("Invalid email address");
+    } else {
+      setEmailError("");
```
**Could be improved**

If it’s interesting for you here is a custom `validation hook`.

You should put it in a separate file  `useFormAndValidation.js` in folder `hooks` and then import it to the component that needs validation:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

And only one line of code enables validation:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()

```

So, you **don’t need** to make functions for handling inputs, storing values, resetting the form. Everything is already done.

gennady-bars commented on src/components/SignupPopup/SignupPopup.jsx:
```diff
@@ -59,18 +105,20 @@ const SignupPopup = ({
           onChange={handleEmailChange}
```
If you have `onChange` or ` onInput` in an input then you need to add attribute `value` to this input.
This is how React controls inputs.

To read more about **controlled components**: https://reactjs.org/docs/forms.html#controlled-components

**Please, correct it everywhere**

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Pavel!

Great job on submitting your project!  


There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

* When I log in and visit the saved articles page, the header doesn't show my name, and the saved items are not displayed https://snipboard.io/goSj19.jpg . When I reload the page, everything appears
* I can save one and the same article a lot of times https://snipboard.io/UGXMKY.jpg . But the criteria say: when clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page. 
* If you have `onChange` or ` onInput` in an input then you need to add attribute `value` to this input.

There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all  comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨


## Number of points: 75

# Summary:
### Performance Criteria

- [] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - **When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.**
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [x] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--2.5--> API error handling:
  - [x] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 75

## Additional comments:
-

flysputnik17 committed:
mid work save

flysputnik17 committed:
after login user data display and add card fix

flysputnik17 committed:
unSaveArticle function fix

flysputnik17 committed:
added missing values

flysputnik17 commented:
-fixed the problem of username not displayed on the button and on the saved news Router after Login 
-fixed the problem of adding the the news right after Login
-fixed the problem of saving the same news card multiple times and added an unsave function to handle deleteing news from the main route
-add missing values

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!

## Number of points: 97
